### ISSUE TITLE
Marching cubes meshing is done incorrectly(?)

### ISSUE THREAD

**Issue by AdamFrisby** [VAST-AI-Research/TripoSR#12]
First up - **big kudos**. Really great work, by far the best I've seen.

I just wanted to point out that the generation of the mesh from the field is being done incorrectly; there's probably an 'off-by-0.5' error somewhere in the generation of the mesh via marching cubes (I haven't had a chance to look myself yet - but will when I have some free time). I _suspect_ the underlying model is just fine.

You can see the result of the sample generation here:
![image](https://github.com/VAST-AI-Research/TripoSR/assets/114041/88fc181e-d49b-43e0-a093-f540d39f91b6)

If you look at the cushion, you will see a terracing effect - this is a sign that the sampling method used in the cubes is duplicating reads of the edges of a 'cell', instead of reading into the neighbourhood correctly (oft caused by chunking); it can also be caused by rounding of coordinates.

This same issue also seems to be affecting the generation of the normals.

**Comment by mr-lab** [comment]
you may be on to something 
![image](https://github.com/VAST-AI-Research/TripoSR/assets/16530175/86593a79-d5de-4c4f-9931-ab8fcbdb16c4)
reading https://github.com/VAST-AI-Research/TripoSR/blob/main/tsr/system.py
line 163 and after will see if the marching cubes not working correctly or simply an issue with the resolution 256
it could be the  threshold as well , will post any updates , i just need to install it .

**Comment by AdamFrisby** [comment]
I've taken a brief look last night and think it might be an issue upstream in the torchmcubes library.

I'm planning on taking a look this weekend and will see if I can resolve. One thing that could be helpful is exporting a 3D array to disk -- that way I can test the output in another mesher.

**Comment by mr-lab** [comment]
kind of  better results 
![image](https://github.com/VAST-AI-Research/TripoSR/assets/16530175/c9a96b25-7ede-43b0-b813-bdf095e70f37)
system.py line 171 
change threshold from 25 to 5 
i think this works for solid objects like an apple if it's a tree then it might 
resolutions wont fix it 512 takes about 21gb vrm and 1024 about 44GB 
when generating a complex object like a tree  best not to go lower then threshold  10 or it will be a big blob 
the torchmcubes or isosurface_helper have something wrong , or it could be that's how it works .
still can get my head around bunch of things scene_codes is what type , what happens in isosurface_helper  ... 
will see what i can do to replace it with a better reconstruction algo  like directly in trimesh or a pointcloud to mesh method 
i just need to find the 3d points 
cheers

**Comment by andybak** [comment]
> simply change model.extract_mesh(scene_codes) to model.extract_mesh(scene_codes, resolution=some_integer).

Any chance this could be exposed in the Gradio UI?

**Comment by andybak** [comment]
> I've taken a brief look last night and think it might be an issue upstream in the torchmcubes library.
> 
> I'm planning on taking a look this weekend and will see if I can resolve. One thing that could be helpful is exporting a 3D array to disk -- that way I can test the output in another mesher.

Any options to export an internal representation would be useful. Meshes are the lowest common denominator and it's interesting to start thinking of ways to get output between different apps and platforms without baking out a mesh.

**Comment by mr-lab** [comment]
> > simply change model.extract_mesh(scene_codes) to model.extract_mesh(scene_codes, resolution=some_integer).
> 
> Any chance this could be exposed in the Gradio UI?

 



here is how to fully expose it in the UI 
add those lines after line 107 in gradio app .py 

                    resolution= gr.Slider(
                        label="resolution of mesh",
                        minimum=128,
                        maximum=512,
                        value=256,
                        step=64,
                    )
                    threshold = gr.Slider(
                        label="threshold  of merging the marching cubes cells or what ever",
                        minimum=1,
                        maximum=100,
                        value=25,
                        step=1,
                    )



then in line  142: inputs=[input_image, do_remove_background, foreground_ratio],

change to : inputs=[input_image, do_remove_background, foreground_ratio,resolution,threshold ],


then in line  60  mesh = model.extract_mesh(scene_codes)[0]
change to : mesh = model.extract_mesh(scene_codes,resolution,threshold )[0]

still wont change the quality of the mesh just how dense it is , will give it a try again perhaps threshold need to be 25 or resolution/2.

will test that when get back to the desk

**Comment by mr-lab** [comment]
> I think it could be due to marching cubes with resolution 256. Could you please try with higher resolutions and see if things get better? To use different resolutions, simply change `model.extract_mesh(scene_codes)` to `model.extract_mesh(scene_codes, resolution=some_integer)`.

![image](https://github.com/VAST-AI-Research/TripoSR/assets/16530175/c910f749-f93b-4a83-9706-f50eb02af461)

resolutions  have no effects . it the cell merging as [AdamFrisby](https://github.com/AdamFrisby) said

**Comment by AdamFrisby** [comment]
Threshold should actually be more like 0. It's probably intended as a basic de-noising mechanism; but it's going to cause weird cutoffs.

It might be worth testing plugging in an alternative de-noiser, like this one: https://github.com/hkuadithya/CUDA-NLML-MRI-Denoising

**Comment by mr-lab** [comment]
> Threshold should actually be more like 0. It's probably intended as a basic de-noising mechanism; but it's going to cause weird cutoffs.
> 
> It might be worth testing plugging in an alternative de-noiser, like this one: https://github.com/hkuadithya/CUDA-NLML-MRI-Denoising

at 0.1 is just a big blob 
![image](https://github.com/VAST-AI-Research/TripoSR/assets/16530175/f3efb152-7f81-4c9a-ba2f-c775d661144d)
it mess up the color vertex as well 
0 is just an error  cannot reshape tensor 
Threshold  is  simply giving a set of points , how far/bad those points are  before we completely disregard them from the final mesh

**Comment by mr-lab** [comment]
> > kind of better results ![image](https://private-user-images.githubusercontent.com/16530175/310329451-c9a96b25-7ede-43b0-b813-bdf095e70f37.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3MzQzMzYsIm5iZiI6MTcwOTczNDAzNiwicGF0aCI6Ii8xNjUzMDE3NS8zMTAzMjk0NTEtYzlhOTZiMjUtN2VkZS00M2IwLWI4MTMtYmRmMDk1ZTcwZjM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDE0MDcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwZDkzNjNjMmU3MDBiYzQwOGU1ZTk1NTVjZWUxMWNhZjFhZGJiNDZjMmFhNTcyYjFjZDI5YWYwZmIyZmZmZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3-qyVOaGTmF0UXuPIEjoiaSQYexnGpcii9qj6FEl6go) system.py line 171 change threshold from 25 to 5 i think this works for solid objects like an apple resolutions wont fix it 512 takes about 21gb vrm and 1024 about 44GB when generating a complex object like a tree best not to go lower then threshold 10 or it will be a big blob the torchmcubes or isosurface_helper have something wrong , or it could be that's how it works . still can get my head around bunch of things scene_codes is what type , what happens in isosurface_helper ... will see what i can do to replace it with a better reconstruction algo like directly in trimesh or a pointcloud to mesh method i just need to find the 3d points cheers
> 
> 1024 at 44GB ?
> 
>  Peek.2024-03-06.14-35.mp4 
> Almost 😉
> 
> **Morbid Interest:** ![image](https://private-user-images.githubusercontent.com/78346668/310547622-2a8c2815-6eb5-417a-8a8f-a925e3085a55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3NTgwOTcsIm5iZiI6MTcwOTc1Nzc5NywicGF0aCI6Ii83ODM0NjY2OC8zMTA1NDc2MjItMmE4YzI4MTUtNmViNS00MTdhLThhOGYtYTkyNWUzMDg1YTU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDIwNDMxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlYTZmMzFmNzYwOTE3YmUxM2QwMzQyNDhhNzY3MTVjOTMzY2YzYTIxYmNjYzJiOWE0YzY3MjZjNjM4NmY1OGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Rm-zMtBYiMYBF0OSVaWSyzEuAujlGzSLz92fawCjENA)
> 
> I don't think the front is as noticeably different other than around the eyes and mouth: ![image](https://private-user-images.githubusercontent.com/78346668/310548249-6ec1a0f1-2d6d-455d-969a-d003f79fe2fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3NTgwOTcsIm5iZiI6MTcwOTc1Nzc5NywicGF0aCI6Ii83ODM0NjY2OC8zMTA1NDgyNDktNmVjMWEwZjEtMmQ2ZC00NTVkLTk2OWEtZDAwM2Y3OWZlMmZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDIwNDMxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5YmZkMDM0YzNkMTc2OTI2ZmFmMWEyYzY5MWE0NzQ4NWE1Yzk2ZmI0ZTM3N2M1ZmJjYzQ1M2M1NjJhMTg1OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gbtUpQLCYgSpj3lMjceMoMkDKwRTtSbZJnbdSKdv3OM)
> 
> **The model files for reference:** [256+1024.zip](https://github.com/VAST-AI-Research/TripoSR/files/14511824/256%2B1024.zip)
> 
> **The source input image as reference:** ![cutecat](https://private-user-images.githubusercontent.com/78346668/310552248-69f2a2f4-9319-4b1d-b90a-0404326d86e6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3NTgwOTcsIm5iZiI6MTcwOTc1Nzc5NywicGF0aCI6Ii83ODM0NjY2OC8zMTA1NTIyNDgtNjlmMmEyZjQtOTMxOS00YjFkLWI5MGEtMDQwNDMyNmQ4NmU2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDIwNDMxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0MjJmZjc3Y2ZmODgwNjVjZGFkNTI5ZmZmMTY4NmFjYzNhNTE2ZGNjYzY2MDNlNWZiY2VhMjBmYmE2MDg2ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PF0VBXZ6P5_Gk4j5gZ5F2zAX8i9w418eVOOjQp9oIs4) Reference: https://civitai.com/images/6932312
> 
> Basic threshold comparison: ![image](https://private-user-images.githubusercontent.com/78346668/310559597-1dbb531e-14de-4f82-8664-e6b41981db45.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3NTgwOTcsIm5iZiI6MTcwOTc1Nzc5NywicGF0aCI6Ii83ODM0NjY2OC8zMTA1NTk1OTctMWRiYjUzMWUtMTRkZS00ZjgyLTg2NjQtZTZiNDE5ODFkYjQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDIwNDMxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyZjBmMzkzZjI1YjQzNGIwMzZmOGU4ZDY2NTVhZGRlZjJjM2ZkMTJhY2Y2MzMzOWYxZTI1MzI4YzU5OWQzMTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lJ_Z8Km2aGghwBpjT3H2XJSwk7mHT-qr4b8S8mLsriQ)

thank you for verifying the resolutions , as for the  threshold  if the subject has multiple sub mesh islands like the hair on that cat or branches on a tree it will be bad plus no correct vertex colors , 10 is the best value 
and use lower values for simple objects like a box apple or a chair , it's a hit or miss thing 

tldr , the 3d points generated from this model is solid gold the mesh generation has an issue .
